**qui**ck **C**lonal **A**nalysis **T**oolkit—`quiCAT`—is a toolkit for extracting, and analyzing clonal populations from sequencing data. Designed for flexibility and performance, it simplifies the process genetic barcodes extraction from bulkDNA, single-cell, and spatial transcriptomics datasets.

---

## Key Features

- **Flexible Input Formats**: Supports FASTQ and BAM file formats, and a wide array of sequencing technologies.
- **Optimized for Speed**: ngrams based barcodes collapsing to speed up the extraction.

<figure>
  <img src="./assets/overview.pdf" alt="Description" />
  <figcaption>Overview of quiCAT workflow.</figcaption>
</figure>

---

## Getting Started

Refer to the [Installation guide](installation.md) to install `quiCAT`

Refer to the [Usage Principles](./usage/usage.md), for an introduction on how to use `quiCAT` modules.

### Contributing

Contributions are welcome! If you would like to contribute to `quiCAT`, please read the [contributing guidelines](contributing.md).

### License

`quiCAT` is licensed under the MIT License.

For more information, visit the [GitHub repository](https://github.com/theislab/quicat/).
